src/main/java/org/littleshoot/proxy/extras/SelfSignedSslEngineSource.java

```diff
@@ -70,6 +98,83 @@ public SelfSignedSslEngineSource() {
+
+        		//modify certificats
+        		Random random = new Random();
+        		remoteServerCertInfo.set(X509CertInfo.VERSION, new CertificateVersion(CertificateVersion.V3));
+        		remoteServerCertInfo.set(X509CertInfo.ISSUER + "." + CertificateSubjectName.DN_NAME, issuer);
+        		remoteServerCertInfo.set(X509CertInfo.KEY, new CertificateX509Key(caPublicKey));
+        		remoteServerCertInfo.set(X509CertInfo.SERIAL_NUMBER, new CertificateSerialNumber(new BigInteger(random.nextLong()+"")));
+        		
+        		
+
+        		X509CertImpl newCert = new X509CertImpl(remoteServerCertInfo);
+        		// The inner and outer signature algorithms have to match.
+        		// The way we achieve that is really ugly, but there seems to be no
+        		// other solution: We first sign the cert, then retrieve the
+        		// outer sigalg and use it to set the inner sigalg
+        		newCert.sign(caPrivateKey, SIG_ALG_NAME);
+        		AlgorithmId sigAlgid = (AlgorithmId)newCert.get(X509CertImpl.SIG_ALG);
+        		remoteServerCertInfo.set(CertificateAlgorithmId.NAME + "." + CertificateAlgorithmId.ALGORITHM, sigAlgid);
+
+        		CertificateExtensions extensions = (CertificateExtensions)remoteServerCertInfo.get(X509CertInfo.EXTENSIONS);
+        		Enumeration<Extension> elements = extensions.getElements();
+        		CertificateExtensions newExtensions = new CertificateExtensions();
+        		while(elements!=null && elements.hasMoreElements()){
+        			Extension ext = elements.nextElement();
+        			if(ext instanceof SubjectAlternativeNameExtension){
+        				newExtensions.set(SubjectAlternativeNameExtension.NAME, ext);
+        			}
+        		}
+
+        		//newExtensions.set(SubjectKeyIdentifierExtension.NAME, new SubjectKeyIdentifierExtension(new KeyIdentifier(caPublicKey).getIdentifier()));
+        		remoteServerCertInfo.set(X509CertInfo.EXTENSIONS, newExtensions);
+
+        		// Sign the new certificate
+        		newCert = new X509CertImpl(remoteServerCertInfo);
+        		newCert.sign(caPrivateKey, SIG_ALG_NAME);
+
+
+        		KeyStore serverKeyStore = KeyStore.getInstance(KEY_STORE_TYPE);
+        		serverKeyStore.load(null, PASSWORD.toCharArray());
+
+        		serverKeyStore.setCertificateEntry(ALIAS, newCert);
+        		serverKeyStore.setKeyEntry(ALIAS, caPrivateKey, PASSWORD.toCharArray(), new Certificate[] { newCert, caCert });
+
+        		final KeyManagerFactory keyManagerFactory = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
+        		keyManagerFactory.init(serverKeyStore, PASSWORD.toCharArray());
+
+        		newSslContext.init(keyManagerFactory.getKeyManagers(), new TrustManager[] { new TrustEveryone() }, null);
+        		
+        		sslContextCache.put(sslContextKey, new SoftReference<SSLContext>(newSslContext));
+    		}
+    		
+    		return newSslContext.createSSLEngine();
+            
+    	}catch(Exception e){
+    		throw new RuntimeException("newSslEngine", e);
```

Comment:
It 's almost impossible that we will throw a exception here, because we have successfully established a SSL connection with the remote server, and certainly  we are able to forge a certificate and return a SslContext instance.